Added Sub method

ctessum opened the pull request.

sbinet commented on badunit/badunit.go:
```diff
@@ -0,0 +1,83 @@
+package badunit
+
+import "github.com/ctessum/unit"
```
this doesn't look right :)

sbinet commented on dimensions.go:
```diff
@@ -1,7 +1,3 @@
-// Copyright ©2013 The gonum Authors. All rights reserved.
```
why is this being removed?

sbinet commented on badunit/badunit.go:
```diff
@@ -0,0 +1,83 @@
+package badunit
```
License header

sbinet commented on unit.go:
```diff
@@ -0,0 +1,338 @@
+
+import "fmt"
+
+// Unit is a type a value with generic SI units. Most useful for
```
`s/Unit is a type a value/Unit describes a value in generic SI units/`
or something along these lines?
(not a native english speaker, but I had issues parsing that sentence)

sbinet commented on unit.go:
```diff
@@ -0,0 +1,338 @@
+// base units struct. The value is always in SI Units.
+//
+// Example: To create an acceleration of 3 m/s^2, one could do
+// myvar := CreateUnit(3.0, &Dimensions{unit.LengthDim: 1, unit.TimeDim: -2})
```
`s/CreateUnit/New/`

sbinet commented on unit.go:
```diff
@@ -0,0 +1,338 @@
+//
+// Example: To create an acceleration of 3 m/s^2, one could do
+// myvar := CreateUnit(3.0, &Dimensions{unit.LengthDim: 1, unit.TimeDim: -2})
+func New(value float64, d Dimensions) *Unit {
```
I surmise this is a leftover from the `Uniter` interface but: why the `*Unit` indirection?
wouldn't returning it by value put less pressure on the GC?

sbinet commented on unit.go:
```diff
@@ -0,0 +1,338 @@
+// myvar := CreateUnit(3.0, &Dimensions{unit.LengthDim: 1, unit.TimeDim: -2})
+func New(value float64, d Dimensions) *Unit {
+	u := &Unit{
+		dimensions: make(map[Dimension]int),
```
`s/make(map[Dimension]int)/make(Dimensions)/`

sbinet commented on unit.go:
```diff
@@ -0,0 +1,338 @@
+// Clone makes a copy of a unit.
+func (u *Unit) Clone() *Unit {
+	o := &Unit{
+		dimensions: make(map[Dimension]int),
```
`s/make(map[Dimension]int)/make(Dimensions)/`

sbinet commented on unit.go:
```diff
@@ -0,0 +1,338 @@
+
+// DimensionsMatch checks if the dimensions of two *Units are the same.
+func DimensionsMatch(a, b *Unit) bool {
+	aUnit := a
```
I don't see the need to introduce `aUnit` nor `bUnit`.

sbinet commented on unit.go:
```diff
@@ -0,0 +1,338 @@
+		return false
+	}
+	for key, val := range d {
+		if d2[key] != val {
```
what if `d2` does not have key `key` but another one that `d` hasn't ?
shouldn't we also use `v2,ok := d2[key]` and/or also range-loop over `d2` ?

sbinet commented on unit.go:
```diff
@@ -0,0 +1,338 @@
+func Sub(u ...*Unit) *Unit {
+	return operatePanic(func(o, uu *Unit, i int) {
+		if !DimensionsMatch(o, uu) {
+			panic(fmt.Errorf("Mismatched dimensions in subtraction: "+
```
prefix with `unit: `

sbinet commented on unit.go:
```diff
@@ -0,0 +1,338 @@
+// Sub subtracts uu from u, modifying u instead of creating a copy.
+func (u *Unit) Sub(uu *Unit) {
+	if !DimensionsMatch(u, uu) {
+		panic(fmt.Errorf("Mismatched dimensions in subtraction: "+
```
prefix with `unit: `

sbinet commented on unit.go:
```diff
@@ -0,0 +1,338 @@
+func Max(u ...*Unit) *Unit {
+	return operateIgnore(func(o, uu *Unit, i int) {
+		if !DimensionsMatch(o, uu) {
+			panic(fmt.Errorf("Mismatched dimensions in addition: "+
```
prefix with `unit: `

sbinet commented on unit.go:
```diff
@@ -0,0 +1,338 @@
+func Min(u ...*Unit) *Unit {
+	return operateIgnore(func(o, uu *Unit, i int) {
+		if !DimensionsMatch(o, uu) {
+			panic(fmt.Errorf("Mismatched dimensions in addition: "+
```
prefix with `unit: `

kortschak commented:
What's the rationale for this PR?

ctessum commented:
I had forgotten about this. I made this pull request over a year ago to just add a "Sub" method for doing subtraction. The rationale for that should be obvious enough.

Since then I have been using this package for a project, and while using it I became frustrated with the combination of statically and dynamically specified units types, and started to feel that because the statically specified types are not able to do multiplication or division, they are pretty much worthless. Static typing is obviously desirable in theory but this implementation wasn't working for me at all and I haven't been able to come up with an elegant solution that involves static typing. The only thing I've thought of is to add a compiler tag to turn on or off dynamic unit checking, but that doesn't seem ideal.

So in my fork I deleted the static types, expanded the dynamic type, and added a "badunit" subpackage for converting from other unit systems into SI. 

Meanwhile, I had forgotten that this PR was still tracking the master branch of my fork. Perhaps the best way forward would be to close this PR (I will try to implement the suggestions regardless), and I can open new PR(s) with the changes in my branch that are deemed worthy of integrating into the main repository. Let me know if any of the changes are of interest for inclusion, and if so which ones. I can start a discussion on the google group if that seems more appropriate.

kortschak commented:
Yes, close this and reopen another with the Sub addition.
